docs/docs/schema-customization.md

```diff
@@ -0,0 +1,765 @@
+to the `AuthorJson.joinedAt` and the `MarkdownRemark.frontmatter.publishedAt`
+fields. Those options are available as field arguments when querying those fields:
+
+```graphql
+query {
+  allAuthorJson {
+    joinedAt(fromNow: true)
+  }
+}
+```
+
+For `publishedAt` we also provide a default `formatString` which will be used
+when no explicit formatting options are provided in the query.
+
+### Setting default field values
+
+For setting default field values, Gatsby currently does not (yet) provide an
+out-of-the-box extension, so resolving a field to a default value (instead of
+`null`) requires manually adding a field resolver. For example, to add a default
+tag to every blog post:
+
+```js:title=gatsby-node.js
+exports.sourceNodes = ({ action, schema }) => {
+  const { createTypes } = actions
+  const typeDefs = [
+    "type MarkdownRemark implements Node { frontmatter: Frontmatter }",
+    schema.buildObjectType({
+      name: "Frontmatter",
+      fields: {
+        tags: {
+          type: "[String!]",
+          resolve(source, args, context, info) {
+            // For a more generic solution, we could pick the field value from
+            // `source[info.fieldName]`
+            const { tags } = source
+            if (source.tags == null || (Array.isArray(tags) && !tags.length)) {
+              return ["uncategorized"]
+            }
+            return tags
+          },
+        },
+      },
+    }),
+  ]
+  createTypes(typeDefs)
+}
+```
+
+# createResolvers API
+
+While it is possible to directly pass `args` and `resolvers` along the type
+definitions using Gatsby Type Builders, an alternative approach specifically
+tailored towards adding custom resolvers to fields is the `createResolvers` Node
+API.
+
+```js:title=gatsby-node.js
+exports.createResolvers = ({ createResolvers }) => {
+  const resolvers = {
+    Frontmatter: {
+      author: {
+        resolve(source, args, context, info) {
+          return context.nodeModel.getNodeById({
+            id: source.author,
+            type: "AuthorJson",
+          })
+        },
+      },
+    },
+  }
+  createResolvers(resolvers)
+}
+```
+
+Note that `createResolvers` allows adding new fields to types, modifying `args`
+and `resolver` -- but not overriding the field type. This is because
+`createResolvers` is run last in schema generation, and modifying a field type
+would mean having to regenerate corresponding input types (`filter`, `sort`),
+which we want to avoid. If possible, specifying field types should be done with
+the `createTypes` action.
+
+## Accessing Gatsby's data store from field resolvers
+
+As mentioned above, Gatsby's internal data store and query capabilities are
+available to custom field resolvers on the `context.nodeModel` argument passed
+to every resolver. Accessing node(s) by `id` (and optional `type`) is possible
+with `getNodeById` and `getNodesByIds`. To get all nodes, or all nodes of a
```

Comment:
Are these API methods defined anywhere else? I'm wondering about API docs for these where each method is in its own section in a repeatable format (in a separate doc?).

Reply:
It's in `/docs/node-model`. Added a link.